src/rgw/rgw_auth.cc

```diff
@@ -853,6 +853,12 @@ void rgw::auth::RoleApplier::modify_request_state(const DoutPrefixProvider *dpp,
 {
   for (auto it: role.role_policies) {
     try {
+      const char *variable = "${aws:username}";
+      std::string str(variable);
+      if (it.find(variable,0) != string::npos){
+        auto length = str.length();
+        it.replace(it.find(variable,0),length,s->user->get_id().id);
```

Comment:
> This example uses the aws:username key, which returns the user's friendly name (like "Adele" or "David"). Under some circumstances, you might want to use the aws:userid key instead, which is a globally unique value.

aws supports both `aws:username` and `aws:userid`, so `s->user->get_id().id` would correspond to the latter. we have `s->user->get_display_name()` for `aws:username`

Reply:
Understood, will update the pull request with the changes.

Reply:
Moved the Policy var substitution to `Statement::eval()`. Are the new changes ok?
Haven't added support for paths or `aws:userid` yet. Wanted to clarify a few doubts first.

In the context of **AssumeRoleWithWebIdentity**, **Keycloak**
Role policy : 
`{"Version":"2012-10-17","Statement":{"Effect":"Allow","Action":"s3:*","Resource":"arn:aws:s3:::${aws:username}"}}`

User  (as per keycloak):

- preferred_username : Bobbet
- user_id : ${keycloak_user_id}

Sts session :

-  RoleSessionName : Bob
- key & secret : none


The following values are made available in the environment.
```
sts:authentication -:- true
aws:username -:- ${keycloak_user_id}  
aws:UserAgent -:- Boto3/1.24.92 Python/3.10.6 Windows/10 Botocore/1.27.92  
aws:SourceIp -:- 192.168.1.3  
aws:PrincipalType -:- User  
aws:CurrentTime -:- 1666118342  
aws:TokenIssueTime -:- 2022-10-18T18:39:01.958083200Z  
aws:EpochTime -:- 2022-10-18T18:39:02.104018926Z  
aws:userid -:- ${ceph_role_id}:Bob  
```

How would the `userid` substitution be handled?
can the following values in the Environment instead be? 
```
aws:userid -:- ${keycloak_user_id}:${RoleSessionName}
aws:username -:- ${keycloak_preferred_username}
ceph:roleId -:- ${ceph_role_id}
```

Reply:
> How would the `userid` substitution be handled? can the following values in the Environment instead be?
> 
> ```
> aws:userid -:- ${keycloak_user_id}:${RoleSessionName}
> aws:username -:- ${keycloak_preferred_username}
> ceph:roleId -:- ${ceph_role_id}
> ```

that's a good question, i'll need a second opinion on this from @pritha-srivastava @mattbenjamin @adamemerson

Reply:
@codeyz @pritha-srivastava @mattbenjamin @adamemerson any updates on the above?

Reply:
We recently ran into EXACTLY this very question: How to create a role policy that is generic for multiple / all users?
Leading to the question about AWS policy variables which are quite common to efficiently create role policies.
So thanks for picking up on that concept and starting on this PR!

> In the context of **AssumeRoleWithWebIdentity**, **Keycloak**
> Role policy :
> `{"Version":"2012-10-17","Statement":{"Effect":"Allow","Action":"s3:*","Resource":"arn:aws:s3:::${aws:username}"}}`

@Yejneshwar this policy example using `${aws:username}` would "only" grant access to a particular bucket with the name equal to the user name right?
How would "ownership" of buckets work then? (How) would a user be able to create a new bucket and then automatically have access to it and the others he owns? Reading the communicated example (https://docs.ceph.com/en/quincy/radosgw/STS/#examples) for a policy `{"Version":"2012-10-17","Statement":{"Effect":"Allow","Action":"s3:*","Resource":"arn:aws:s3:::*"}}` reads to me as giving the role FULL access to any bucket, regardless of the ownership?

Without STS / assumed roles RADOSGW does apply a filter via the owner / user accessing. Any each bucket does have an owner and also a tenant. How does this work with assumed roles? Or can I assume the role of the owner somehow?

Reply:
> > How would "ownership" of buckets work then? (How) would a user be able to create a new bucket and then automatically have access to it and the others he owns?
> 
> To my understanding, having such a policy will definitely restrict a user's access to one single bucket; named after their username. AWS doesn't allow for paths in the bucket name itself. Hence for the more general use case, example policies [here](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_variables.html) may be more suitable (this will require support for default values, substitutions in paths, and, in strings [ please see this example ](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_variables.html#:~:text=For%20example%2C%20if%20a,for%20joining%20a%20team.)).

I am still confused about the isolation the ownership (potentially) imposes. Reading https://docs.ceph.com/en/quincy/radosgw/STS/#examples

> The following is an example of the AssumeRole API call, which shows steps to create a role, assign a policy to it (that allows access to S3 resources), assuming a role to get temporary credentials and accessing S3 resources using those credentials. In this example, TESTER1 assumes a role created by TESTER, to access S3 resources owned by TESTER, according to the permission policy attached to the role.

it seems to imply that "below" the role policy there is the ownership. In clear, TEST can only grant access to the resource he owns via any policy. 

> **In this example, TESTER1 assumes a role created by TESTER, to access S3 resources owned by TESTER, according to the permission policy attached to the role.**

^^  @cbodley can you clarify?

Reply:
Thats right, if TESTER1 assumes a role created by TESTER, access is restricted to the buckets owned by TESTER only.

Reply:
Please keep in mind this is the case for AssumeRole due to the policy `'''{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":["arn:aws:iam:::user/TESTER1"]},"Action":["sts:AssumeRole"]}]}'''`.
Policy is different for AssumeRoleWithWebIdentity.

Reply:
I took my questions to a thread on the ML: https://lists.ceph.io/hyperkitty/list/ceph-users@ceph.io/thread/NCFJF7TFPNU4PMX7TFOJ5OEKKQSSQXV5/

Maybe you care to discuss (and potentially explain) these matters some more?